src/analysis/ast_dependency_detector.rs

```diff
@@ -0,0 +1,974 @@
+        Vec<(&'a QualifiedContractIdentifier, &'a [SymbolicExpression])>,
+    >,
+    pending_trait_checks: HashMap<
+        // trait that is not yet defined
+        &'a TraitIdentifier,
+        // list of contracts that need to be checked once this trait is
+        // defined, together with the function called and the associated args.
+        Vec<(
+            &'a QualifiedContractIdentifier,
+            &'a ClarityName,
+            &'a [SymbolicExpression],
+        )>,
+    >,
+    params: Option<Vec<TypedVar<'a>>>,
+    preloaded: &'a BTreeMap<QualifiedContractIdentifier, ContractAST>,
+}
+
+impl<'a> ASTDependencyDetector<'a> {
+    pub fn detect_dependencies(
+        contract_asts: &'a HashMap<QualifiedContractIdentifier, ContractAST>,
+        preloaded: &'a BTreeMap<QualifiedContractIdentifier, ContractAST>,
+    ) -> HashMap<QualifiedContractIdentifier, HashSet<QualifiedContractIdentifier>> {
+        let mut detector = Self {
+            dependencies: HashMap::new(),
+            current_contract: None,
+            defined_functions: HashMap::new(),
+            defined_traits: HashMap::new(),
+            pending_function_checks: HashMap::new(),
+            pending_trait_checks: HashMap::new(),
+            params: None,
+            preloaded,
+        };
+
+        let mut preloaded_visitor = PreloadedVisitor {
+            detector: &mut detector,
+            current_contract: None,
+        };
+
+        for (contract_identifier, ast) in preloaded {
+            preloaded_visitor.current_contract = Some(contract_identifier);
+            traverse(&mut preloaded_visitor, &ast.expressions);
+        }
+
+        for (contract_identifier, ast) in contract_asts {
+            detector
+                .dependencies
+                .insert(contract_identifier.clone(), HashSet::new());
+            detector.current_contract = Some(contract_identifier);
+            traverse(&mut detector, &ast.expressions);
+        }
+
+        detector.dependencies
+    }
+
+    pub fn order_contracts(
+        dependencies: &HashMap<QualifiedContractIdentifier, HashSet<QualifiedContractIdentifier>>,
+    ) -> Vec<&QualifiedContractIdentifier> {
+        let mut lookup = BTreeMap::new();
+        let mut reverse_lookup = Vec::new();
+
+        let mut index: usize = 0;
+
+        if dependencies.is_empty() {
+            return vec![];
+        }
+
+        for (contract, _) in dependencies {
+            lookup.insert(contract, index);
+            reverse_lookup.push(contract);
+            index += 1;
+        }
+
+        let mut graph = Graph::new();
+        for (contract, contract_dependencies) in dependencies {
+            let contract_id = lookup.get(contract).unwrap();
+            graph.add_node(*contract_id);
+            for dep in contract_dependencies.iter() {
+                let dep_id = match lookup.get(dep) {
+                    Some(id) => id,
+                    None => {
+                        println!(
```

Comment:
When using the repl and clarinet as a library is a desktop app or a server, these `process::exit` are pretty incovenient.
I'm working on getting rid of these pattern across the board, could we return an error here instead?

Reply:
Note: if these code paths are not reachable you can use the macro `unreachable!()`

Reply:
This case is reachable if there is an unmet dependency, so I think returning a `Result` probably makes sense here.

Reply:
Addressed with the last commit.